packages/typespec-client-generator-core/test/decorators/legacy-hierarchy-building.test.ts

```diff
@@ -0,0 +1,535 @@
+      @service
+      namespace TestService;
+
+      model A extends B {
+        propA: string;
+      }
+
+      @Legacy.legacyHierarchyBuilding(A)
+      model B extends C {
+        propB: string;
+      }
+
+      @usage(Usage.input)
+      model C {
+        propC: string;
+      }
+
+      @route("/test")
+      op test(): A;
+    `);
+
+  const modelA = runner.context.sdkPackage.models.find((m) => m.name === "A");
+  const modelB = runner.context.sdkPackage.models.find((m) => m.name === "B");
+  const modelC = runner.context.sdkPackage.models.find((m) => m.name === "C");
+
+  ok(modelA);
+  ok(modelB);
+  ok(modelC);
+
+  strictEqual(modelA.baseModel?.name, "B");
+  strictEqual(modelB.baseModel?.name, "A");
+  strictEqual(modelC.baseModel, undefined);
+
+  strictEqual(modelA.properties.length, 1);
+  strictEqual(modelA.properties[0].name, "propA");
+  strictEqual(modelB.properties.length, 1);
+  strictEqual(modelB.properties[0].name, "propB");
+  strictEqual(modelC.properties.length, 1);
+  strictEqual(modelC.properties[0].name, "propC");
+});
+
+it("conflicting inheritance", async () => {
+  const [_, diagnostics] = await runner.compileAndDiagnose(`
+      @service
+      namespace TestService;
+
+      model A {
+        propA: string;
+      }
+
+      model B {
+        propB: string;
+      }
+
+      @Legacy.legacyHierarchyBuilding(B)
+      model C extends A {  // Doesn't have propB but @Legacy.legacyHierarchyBuilding says B
+        propC: string;
+      }
+
+      @route("/test")
+      op test(): C;
+    `);
+
+  // Should warn about conflicting inheritance
+  expectDiagnostics(diagnostics, {
+    code: "@azure-tools/typespec-client-generator-core/legacy-hierarchy-building-conflict",
+  });
+});
+
+it("inheritance override with template models", async () => {
+  await runner.compileWithBuiltInService(`
+      @discriminator("type")
+      model Container<T> {
+        type: string;
+        data: T;
+      }
+
+      model StringContainer extends Container<string> {
+        type: "string";
+      }
+
+      @Legacy.legacyHierarchyBuilding(StringContainer)
+      model SpecialContainer extends Container<string> {
+        type: "special";
+        metadata: Record<string>;
+      }
+
+      @route("/containers")
+      op getContainer(): Container<string>;
+    `);
+
+  const models = runner.context.sdkPackage.models;
+  const specialContainerModel = models.find((m) => m.name === "SpecialContainer");
+
+  ok(specialContainerModel);
+  // Should inherit from StringContainer instead of Container<string>
+  strictEqual(specialContainerModel.baseModel?.name, "StringContainer");
+});
+
+it("without polymorphism", async () => {
+  await runner.compileWithBuiltInService(`
+      model A {
+        kind: string;
+      }
+
+      alias BContent = {
+        foo: string;
+      };
+
+      model B extends A {
+        kind: "B";
+        ...BContent;
+      }
+
+      @Legacy.legacyHierarchyBuilding(B)
+      @usage(Usage.input)
```

Comment:
why usage?

Reply:
changed to add an op to prevent miss-leading.